### PR TITLE
Add KSQL support for BETWEEN in predicates #1004.

### PR DESCRIPTION
**Pull request by agavra**
This is standard SQL that users would expect to be supported, which
makes queries that compare ranges easier to read. See #1004 

### Description 

Introduce support for BETWEEN, for queries such as 
```
SELECT * FROM topic WHERE field BETWEEN valA AND valB
```

The BETWEEN operation follows convention in that the range is *inclusive* of the endpoints. Any `null` values will immediately fail the between range (in either the field or either of the endpoints). Similar to comparisons, either literals or field references may be used in the limits.

*Open Question:* Should we support comparisons with dates?

### Testing done 
- Unit testing
- Integration testing with `between.json`

### Reviewer checklist
- [ ] Ensure docs are updated if necessary. (eg. if a user visible feature is being added or changed).
- [ ] Ensure relevant issues are linked (description should include text like "Fixes #<issue number>")

### CODE REVIEW

File: docs/developer-guide/syntax-reference.rst [hunk 1]
```diff
@@ -839,7 +839,9 @@ In the above statements from_item is one of the following:
 -  ``from_item LEFT JOIN from_item ON join_condition``
 
 The WHERE clause can refer to any column defined for a stream or table,
-including the two implicit columns ``ROWTIME`` and ``ROWKEY``.
+including the two implicit columns ``ROWTIME`` and ``ROWKEY``. If those
```
Comment (JimGalasyn) [inline docs/developer-guide/syntax-reference.rst]:
```suggestion
including the two implicit columns ``ROWTIME`` and ``ROWKEY``. If these
```

File: docs/developer-guide/syntax-reference.rst [hunk 2]
```diff
@@ -839,7 +839,9 @@ In the above statements from_item is one of the following:
...
+columns contain Strings or Scalar values, the BETWEEN keyword can be used
```
Comment (JimGalasyn) [inline docs/developer-guide/syntax-reference.rst]:
```suggestion
columns contain strings or scalar values, you can use the BETWEEN keyword
```

Comment (agavra) [reply docs/developer-guide/syntax-reference.rst]: Passive tense is a bad habit of mine! Thanks for checking it :)

File: ksql-engine/src/main/java/io/confluent/ksql/codegen/SqlToJavaVisitor.java [hunk 3]
```diff
@@ -574,7 +576,42 @@ private String visitBooleanComparisonExpression(final ComparisonExpression.Type
...
+          break;
+        case MAP:
+          throw new KsqlException("Cannot execute between with MAP values");
+        case ARRAY:
+          throw new KsqlException("Cannot execute between with ARRAY values");
+        case BOOLEAN:
+          throw new KsqlException("Cannot execute between with BOOLEAN values");
+        default:
+          expression += "{min} <= {value} && {value} <= {max}";
+          break;
```
Comment (spena) [inline ksql-engine/src/main/java/io/confluent/ksql/codegen/SqlToJavaVisitor.java]: There is a BYTES type as well. is this the way to evaluate a BYTES column? 
Btw, wouldn't be better to use a default case to throw an exception in case a new type is added in the future? Like DATE, TIME, etc. That way we could throw the error as not supported instead of a bug because those types do not accept the <= evaluation.

Comment (agavra) [reply ksql-engine/src/main/java/io/confluent/ksql/codegen/SqlToJavaVisitor.java]: Actually I misunderstood what BYTES is - I'm not sure there's a good way to compare them by default (it falls more into the ARRAY category than scalar). Any suggestions? Also do we support it? I couldn't find any usage in the Visitor that alludes to BYTES.

Comment (agavra) [reply ksql-engine/src/main/java/io/confluent/ksql/codegen/SqlToJavaVisitor.java]: Actually, this test in `SchemaUtilTest` implies that we don't support it:
```
  @Test(expected = IllegalArgumentException.class)
  public void shouldThrowOnUnknownSchemaType() {
    SchemaUtil.getSchemaTypeAsSqlType(Schema.Type.BYTES);
  }
```

Comment (spena) [reply ksql-engine/src/main/java/io/confluent/ksql/codegen/SqlToJavaVisitor.java]: ok, so we shouldn't use it in the between evaluator then :)

File: ksql-engine/src/main/java/io/confluent/ksql/codegen/SqlToJavaVisitor.java [hunk 4]
```diff
@@ -574,7 +576,44 @@ private String visitBooleanComparisonExpression(final ComparisonExpression.Type
...
+        case INT8:
+        case INT16:
+        case INT32:
+          expression += "{min} <= {value} && {value} <= {max}";
+          break;
+        case STRING:
+          expression += "({value}.compareTo({min}) >= 0 && {value}.compareTo({max}) <= 0)";
+          break;
+        default:
+          throw new KsqlException("Cannot execute between with " + type + " values");
```
Comment (hjafarpour) [inline ksql-engine/src/main/java/io/confluent/ksql/codegen/SqlToJavaVisitor.java]: Maybe change it to `BETWEEN`.

File: ksql-engine/src/test/resources/query-validation-tests/between.json [hunk 5]
```diff
@@ -0,0 +1,147 @@
...
+        {"topic": "OUTPUT", "key": 1, "value": {"THING": 1}, "timestamp": 0},
+        {"topic": "OUTPUT", "key": 5, "value": {"THING": 5}, "timestamp": 0}
+      ]
+    },
+    {
+      "name": "test BETWEEN with integers and variable values",
+      "format": ["JSON"],
+      "statements": [
+        "CREATE STREAM TEST (source int, min int, max int) WITH (kafka_topic='test_topic', value_format='JSON');",
+        "CREATE STREAM OUTPUT AS SELECT source AS THING FROM TEST WHERE source BETWEEN min AND max;"
```
Comment (hjafarpour) [inline ksql-engine/src/test/resources/query-validation-tests/between.json]: Maybe add a test with expressions too. For instance, `source*2 BETWEEN min+2 AND max-min` or something similar.

File: docs/developer-guide/syntax-reference.rst [hunk 6]
```diff
@@ -839,7 +839,8 @@ In the above statements from_item is one of the following:
 -  ``from_item LEFT JOIN from_item ON join_condition``
 
 The WHERE clause can refer to any column defined for a stream or table,
-including the two implicit columns ``ROWTIME`` and ``ROWKEY``.
+including the two implicit columns ``ROWTIME`` and ``ROWKEY``. If these
```
Comment (hjafarpour) [inline docs/developer-guide/syntax-reference.rst]: You don't need to mention this. They can be used as other columns in all operators including BETWEEN.

Comment (agavra) [reply docs/developer-guide/syntax-reference.rst]: Got it - I will add a dedicated section as discussed offline about BETWEEN next to LIKE.

File: docs/developer-guide/syntax-reference.rst [hunk 7]
```diff
@@ -959,6 +959,27 @@ Example:
...
+~~~~~~~
+
+**Synopsis**
+
+.. code:: sql
+
+    WHERE expression [NOT] BETWEEN start_expression AND end_expression;
+
+The BETWEEN operator is used to indicate that a certain value must lie within
+(inclusive of boundaries) a specified range. Currently KSQL supports numeric
```
Comment (JimGalasyn) [inline docs/developer-guide/syntax-reference.rst]:
```suggestion
a specified range, inclusive of boundaries. Currently, KSQL supports numeric
```

File: ksql-engine/src/test/resources/query-validation-tests/between.json [hunk 8]
```diff
@@ -0,0 +1,165 @@
+{
+  "comments": [
```
Comment (big-andy-coates) [inline ksql-engine/src/test/resources/query-validation-tests/between.json]: consider adding test cases to cover _unsupported_ operations, e.g. I'm assuming it doesn't support function calls in the bound expressions. The test can ensure the error message is as expected. (You can define `expectedException`).

File: docs/developer-guide/syntax-reference.rst [hunk 9]
```diff
@@ -959,6 +959,27 @@ Example:
...
+BETWEEN
+~~~~~~~
+
+**Synopsis**
+
+.. code:: sql
+
+    WHERE expression [NOT] BETWEEN start_expression AND end_expression;
+
+The BETWEEN operator is used to indicate that a certain value must lie within
```
Comment (big-andy-coates) [inline docs/developer-guide/syntax-reference.rst]: > Currently, KSQL supports numeric and string values for comparison.

Probably worth fleshing this out a bit. eg. arithmetic is supported, so is struct, array, map access. Anything else? Function calls aren't, anything else?  This probably also highlight any missing JSON tests, both negative and positive.

Comment (agavra) [reply docs/developer-guide/syntax-reference.rst]: Totally agree. I will rephrase this to: "Currently, KSQL supports any expression that resolves to a numeric or string value for comparison". Does that cover things?

File: ksql-engine/src/test/resources/query-validation-tests/between.json (hunk 8 above)
Comment (agavra) [reply ksql-engine/src/test/resources/query-validation-tests/between.json]: For now I'll leave the tests in CodeGenRunnerTest - I can open a new PR with more tests if you feel strongly about adding them to the JSON suite (which tend to run a little longer, which is why I'm erring toward unit tests)

File: ksql-engine/src/main/java/io/confluent/ksql/codegen/SqlToJavaVisitor.java (hunk 3 above)
Comment (apurvam) [reply ksql-engine/src/main/java/io/confluent/ksql/codegen/SqlToJavaVisitor.java]: adding bytes is tracked here: #1282

**Comment by rmoff** [comment]
Thanks @agavra! Two quick questions: 
1. What was the outcome on `Open Question: Should we support comparisons with dates?` - will we support this? 
2. Will this function be in the 5.2 release? 
thanks

**Comment by agavra** [comment]
@rmoff 

1. is still unresolved - I wanted to mirror the support for comparison operators. If we need dates as a first class citizen, I can work on a separate PR to add it to both.
2. Unsure about the release schedule (this is my first PR) @apurvam can you comment?

**Comment by apurvam** [comment]
Yes it will be in the 5.2 release: it was committed to master before the feature branch was cut.
